### DISCUSSION TITLE
Top-level await support

### DISCUSSION THREAD

**Discussion by Berhart** [tauri-apps/tauri#9795] [Q&A] [answered]
Hi all, I'm using Tauri v1.6.5 currently in a professional environment

After some debugging with the help of `tauri build --debug` I noticed on the devtools console that this error comes up
`Unhandled Promise Rejection: ReferenceError: Cannot access uninitialized variable.` and basically the app just shows a blank page. 
Went playing with my code and reached the conclusion that the presence of a [top-level await](https://github.com/tc39/proposal-top-level-await) I'm doing to read an external file is causing this problem.

Am I missing something on the tauri configuration? Isn't it supported yet? Is there a expectation/timeline for it to be available?

Appreciate any reply

**Comment by FabianLars** [comment]
I assume you're on an older macos version? Asking because if we check <https://caniuse.com/mdn-javascript_operators_await_top_level> and <https://tauri.app/v1/references/webview-versions/#macos--ios> shows that it should be available on macos v12 and above.

chromium support is old enough that it should work on all windows versions.

> Is there a expectation/timeline for it to be available?

This is the drawback of using system webviews, we cannot influence this.

**Reply by Berhart** [reply]
Hi @FabianLars, thanks for reaching out 

I'm currently using a macOS Sonoma 14.4 and yeah it's really unfortunate because if the app is launched on Chrome it just works as expected

In case you or someone is willing to help me further on this endeavour, I'm going to leave here a repo that can easily reproduce this experience of mine: https://github.com/Berhart/top-level-await-sveltekit-tauri-example

Basically I have a `config.ts` file that reads a `config.json` either from Sveltekit's static folder (web app) or as a Tauri's external resource (tauri app).

If the app is served on the Browser this is the expected result:

![image](https://github.com/tauri-apps/tauri/assets/36551673/0125b7f5-b9f7-435a-9866-f6dd7dfd9724)

Otherwise, if served as a Tauri app, the result should be the same as above but instead I get:

<img width="801" alt="image" src="https://github.com/tauri-apps/tauri/assets/36551673/dc604f45-3cee-4c97-bd47-9c304ca71771">

**Reply by FabianLars** [reply, marked as answer]
hmm, that's annoying. I would not have expect webkit to have such an annoying bug for this long but assuming this https://bugs.webkit.org/show_bug.cgi?id=242740 is the issue then they are at least working on it quite actively https://github.com/WebKit/WebKit/pull/24122

**Reply by muwoo** [reply]
@Berhart I don’t think this has anything to do with Tauri itself, nor does it have anything to do with webview. I see that you are using `vite` to build, so you need to introduce a vite plugin like [vite-plugin-top-level-await](https://github.com/Menci/vite-plugin-top-level-await) so that vite can be used when building. Can handle `top-level await` correctly.

`top-level await` is only supported by some specific latest browsers, and there will be some compatibility issues, and Vite will not convert it when packaging by default, so this error is reported.

**Reply by Berhart** [reply]
@muwoo thanks for your help. I went trying your suggested solution and the behaviour is still the same.
You're correct, as according to Fabian, it has no direct relation with Tauri as it seems to be a WebKit bug.

It's definitely something non-deterministic, probably some race condition as sometimes it works, sometimes it doesn't 🙃
